### PR TITLE
feat: added L3_DEFAULT, backend_services, and connection tracking (fixes #104)

### LINKED ISSUE #104: Add L3_Default option for External LB to allow all ports and protocols to be forwarded
### Is your feature request related to a problem?

To ease the number of changes to the LB configuration L3_DEFAULT allows all ports and protocols to be forwarded to the backend service. 

### Describe the solution you'd like

Add this option to ip_protocol, and associated backend_service resources

### Describe alternatives you've considered

Considered creating a new LB_EXTERNAL module for this, but probably makes the most sense to merge the functionality in the existing one.

### Additional context

Without this you constantly need to keep adding services to the LB

### PR DESCRIPTION
**Pull request by salsop**
## Description

Added L3_DEFAULT
Added Backend_Service resources to support this.
Added Connection Tracking, as this will be required later for Active/Passive deployments.

## Motivation and Context

Without this fix, you have to specify a large number of Forwarding Rules, this makes it simple to follow everything.

## How Has This Been Tested?

I've tested this in my own lab environment

## Types of changes

- New feature (non-breaking change which adds functionality)

## Checklist

- [x ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes if appropriate.
- [x] All new and existing tests passed.

### CODE REVIEW

File: modules/lb_external/main.tf [hunk 1]
```diff
@@ -4,6 +4,21 @@ terraform {
...
 }
 
+data "google_client_config" "this" {}
+
+locals {
+  # If we were told an exact region, use it, otherwise fall back to a client-default region
+  region = coalesce(var.region, data.google_client_config.this.region)
+
+  # Check for `L3_DEFAULT` as this requires a `google_compute_region_backend_service` resource and `google_compute_region_health_check` health check
+  backend_service_needed = contains([for v in values(var.rules) : lookup(v, "ip_protocol", null)], "L3_DEFAULT")
```
Comment (migara) [inline modules/lb_external/main.tf:14]:
```suggestion
  backend_service_needed = contains([for k, v in var.rules : lookup(v, "ip_protocol", null)], "L3_DEFAULT")
```

Comment (migara) [reply modules/lb_external/main.tf:14]: You can just get rid of the `values` function

File: modules/lb_external/main.tf [hunk 2]
```diff
@@ -4,6 +4,21 @@ terraform {
...
+
+  # Check for protocols that require a `google_compute_target_pool` backend and `google_compute_http_health_check` health check
+  target_pool_protocols = ["TCP", "UDP", "ESP", "AH", "SCTP", "ICMP"]
+  target_pool_needed    = contains([for v in values(var.rules) : contains(local.target_pool_protocols, lookup(v, "ip_protocol", "TCP"))], true)
```
Comment (migara) [inline modules/lb_external/main.tf:18]:
```suggestion
  target_pool_needed    = contains([for k, v in var.rules : contains(local.target_pool_protocols, lookup(v, "ip_protocol", "TCP"))], true)
```

**Review by migara — APPROVED** [review]
Great work @salsop 👏🏽  I have added some suggestions, please accept/reject as needed.

Did you have a chance to check the new inputs/outputs break the existing examples which use this module?

File: modules/lb_external/outputs.tf [hunk 3]
```diff
@@ -10,10 +10,15 @@ output "ip_addresses" {
...
   description = "The self-link of the target pool."
-  value       = google_compute_target_pool.this.self_link
+  value       = try(google_compute_target_pool.this[0].self_link, null)
 }
 
-output "created_health_check" {
+output "created_google_compute_http_health_check" {
   description = "The created health check resource. Null if `create_health_check` option was false."
-  value       = var.create_health_check ? google_compute_http_health_check.this[0] : null
+  value       = var.create_health_check ? try(google_compute_http_health_check.this[0], null) : null
```
[the remark this answers is not in the document]
Comment (salsop) [reply modules/lb_external/outputs.tf]: Changed this over too.

**Comment by salsop** [comment]
> Great work @salsop 👏🏽 I have added some suggestions, please accept/reject as needed.
> 
> Did you have a chance to check the new inputs/outputs break the existing examples which use this module?

I've tested with the `hub_spoke_common` example and all worked well with the changes, so should be a non-breaking change.

**Pull request merged**